lxd/storage/backend_lxd.go

```diff
@@ -2223,6 +2225,225 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
+	if err != nil {
+		return err
+	}
+
+	revert.Add(func() { _ = VolumeDBDelete(b, inst.Project().Name, inst.Name(), volType) })
+
+	// Generate the effective root device volume for instance.
+	err = b.applyInstanceRootDiskOverrides(inst, &vol)
+	if err != nil {
+		return err
+	}
+
+	// Override args.Name and args.Config to ensure volume is created based on instance.
+	args.Config = vol.Config()
+	args.Name = inst.Name()
+
+	if len(args.ConversionOptions) == 0 {
+		// If migration header supplies a volume size, then use that as block volume size instead of pool default.
+		// This way if the volume being received is larger than the pool default size, the block volume created
+		// will still be able to accommodate it.
+		if args.VolumeSize > 0 && contentType == drivers.ContentTypeBlock {
+			l.Debug("Setting volume size from offer header", logger.Ctx{"size": args.VolumeSize})
+			args.Config["size"] = fmt.Sprintf("%d", args.VolumeSize)
+		} else if args.Config["size"] != "" {
+			l.Debug("Using volume size from root disk config", logger.Ctx{"size": args.Config["size"]})
+		}
+
+		volCopy := drivers.NewVolumeCopy(vol)
+
+		err = b.driver.CreateVolumeFromMigration(volCopy, conn, args, nil, op)
+		if err != nil {
+			return err
+		}
+	} else {
+		var imgUploadPath string
+
+		// When conversion option "format" is enabled, we need to upload the image to a temporary location
+		// before converting it into the desired format. The conversion cannot be done in-place, therefore
+		// the image has to be saved in an intermediate location.
+		if slices.Contains(args.ConversionOptions, "format") {
+			// Generate a unique filename for the uploaded image and use the backups directory
+			// as the default destination.
+			conversionID := fmt.Sprintf("conversion_%s_%s", inst.Project().Name, inst.Name())
+			imgUploadPath = filepath.Join(shared.VarPath("backups"), conversionID)
+
+			// Create temporary file where image is uploaded.
+			file, err := os.Create(imgUploadPath)
+			if err != nil {
+				return fmt.Errorf("Failed to create temporary file (%s): %w", imgUploadPath, err)
+			}
+
+			// Remove uploaded image regardless of image conversion success.
+			defer file.Close()
+			defer os.Remove(imgUploadPath)
+		}
+
+		args.ConversionFilePath = imgUploadPath
+
+		// Receive/upload image.
+		err = b.driver.CreateVolumeFromMigration(drivers.NewVolumeCopy(vol), conn, args, nil, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { _ = b.DeleteInstance(inst, op) })
+
+		// Mount volume and get disk path.
+		err = b.driver.MountVolume(vol, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { _, _ = b.driver.UnmountVolume(vol, false, op) })
+		diskPath, err := b.driver.GetVolumeDiskPath(vol)
+		if err != nil {
+			return fmt.Errorf("Failed getting instance volume disk path: %v", err)
+		}
+
+		// If imgUploadPath is empty, it means that the image was uploaded directly into the instance
```

Comment:
If the image was uploaded directly into the instance volume then why are we still doing the checks below with qemu-img?